File: P1.3/P1.3_obligatorio/src/Target.cpp
#include <random>
#include <iostream>
#include "Target.h"
#include "color.h"



Target::Target(double precio)
{
    this->precio = precio;
    this->state = NEUTRO;
    this->demanda = 0;
}

void Target::setPrecio(double p)
{
    this->precio = p;
}

double Target::getPrecio() const
{
    return this->precio;
}

estado Target::getEstado() const
{
    return state;
}

void Target::addLog(std::string log)
{
    logger.push(log);
}

void Target::leerLog()
{
    int c = 0;
    std::cout << YEL(CURS("..logger........")) << std::endl;
    while (!logger.empty())
    {
        std::cout << KBLU << (c + 1) << ": " << logger.front() << RST << std::endl;
        logger.pop();
        c++;
    }
    std::cout << YEL(CURS("........................")) << std::endl;
}

void Target::setDemanda(double d) {
    demanda = d;
}

double Target::getDemanda() const {
    return demanda;
}

string estados[]= {
    "NEUTRO",
    "EN_OFERTA",
    "SIN_OFERTA"
};

void Target::execute() {
    this->leerLog();
    std::cout << BOLD("PRECIO FINAL: ") << precio << " €" << std::endl;
    std::cout << BOLD("ESATDO      : ") << estados[state] << std::endl;
    std::cout << BOLD("DEMANDA     : ") << demanda << " %" << std::endl;

    // Cambiamos estado
    if(precio >= 100)     state = EN_OFERTA;
    else if(precio < 50)  state = SIN_OFERTA;
    else                  state = NEUTRO;


    cout << endl << endl;
}

File: P1.3/P1.3_opcional/Target.dart
import "dart:collection";
import 'Estado.dart';

class Target {
  late double _precio;
  late double _demanda;
  late Estado _state;
  late Queue<String> _logger;

  Target(this._precio) {
    this._demanda = 0;
    this._state = Estado.NEUTRO;
    this._logger = Queue();
  }

  double get precio => _precio;
  Estado get estado => _state;
  double get demanda => _demanda;

  set demanda(double demanda) => _demanda = demanda;
  set precio(double precio) => _precio = precio;

  addLog(String log) => _logger.add(log);

  leerLog() {
    int c = 1;
    String blue = "\x1B[34m";
    String rst = "\x1B[0m";
    print("..Logger.......................");
    while (_logger.isNotEmpty) {
      String frase = _logger.removeFirst();
      print("$blue $c: $frase $rst");
      c += 1;
    }
    print("..............................");
  }

  execute() {
    this.leerLog();
    print("PRECIO FINAL: ${_precio.toStringAsFixed(4)} €");
    print("ESTADO      : ${estados[_state.index]}");
    print("DEMANDA     : ${_demanda.toStringAsFixed(4)} %");

    if (_precio >= 100)
      _state = Estado.EN_OFERTA;
    else if (_precio < 50)
      _state = Estado.SIN_OFERTA;
    else
      _state = Estado.NEUTRO;

    print("\n\n ");
  }
}
